### PR TITLE
gamm: query to get pools with specified parameters 

### PR DESCRIPTION
**Pull request by RusAkh**
Closes: #2906 

## What is the purpose of the change

adds query to get pools with specified parameters

## Testing and Verifying

added unit test

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? no

### CODE REVIEW

**Comment by RusAkh** [comment]
hello @mattverse. Currently this query supports only 4 "filters": min_liquidity, includes_denom, num_tokens, pool_type. If you think some other things might be added here I would love to hear about them. Or if there is anything in my code worth additional work

**Comment by mattverse** [comment]
Hey @RusAkh Sounds awesome, I don't think it's worth supporting num_tokens for now. 
I'm also curious how the implementation of min_liquidity would work, since the chain itself does not know the actual prices, just the relatvie spot price. 

Also lmk or feel free to go ahead more options for the filters that you would like to add!

**Comment by RusAkh** [comment]
@mattverse currently min_liquidity is used in a format of sdk coins such that an amount of every token in pool should be greater or equal than each sdk coin in min_liquidity request. I thought about using an actual price for liquidity in this query (for example 1 mil $) but do not really know how to implement this. What do you think should be done here?

**Comment by mattverse** [comment]
> @mattverse currently min_liquidity is used in a format of sdk coins such that an amount of every token in pool should be greater or equal than each sdk coin in min_liquidity request. I thought about using an actual price for liquidity in this query (for example 1 mil $) but do not really know how to implement this. What do you think should be done here?

So I don't thimk the sdk coin amt represents liquidity. I propose we drop the min_liquidity feature here

**Comment by RusAkh** [comment]
> > @mattverse currently min_liquidity is used in a format of sdk coins such that an amount of every token in pool should be greater or equal than each sdk coin in min_liquidity request. I thought about using an actual price for liquidity in this query (for example 1 mil $) but do not really know how to implement this. What do you think should be done here?
> 
> So I don't thimk the sdk coin amt represents liquidity. I propose we drop the min_liquidity feature here

Hello @mattverse. I currently spoke with other developers and we currently think only two "filters" namely includes_denom and pool_type is enough for now. Just want to make sure it is ok with you

**Comment by RusAkh** [comment]
actually I did not really understand why do not we use a filter for a minimum number of tokens. I think this could actually replace includes_denom filter with something like minimum_amount_of_asset, which will basically be an sdk.coin, which will allow us to query pools with specific asset (includes_denom is basically already doing this), but also with a minimum number of this asset in a pool. If I am wrong, please, I am open to discussion :)

**Comment by mattverse** [comment]
> actually I did not really understand why do not we use a filter for a minimum number of tokens. I think this could actually replace includes_denom filter with something like minimum_amount_of_asset, which will basically be an sdk.coin, which will allow us to query pools with specific asset (includes_denom is basically already doing this), but also with a minimum number of this asset in a pool. If I am wrong, please, I am open to discussion :)

Oh I see, the feature ur suggesting does not support a single filter, but multiple filters at once, 

good point, I am very supportive of this

**Comment by RusAkh** [comment]
hello @mattverse done

File: x/gamm/keeper/grpc_query.go [hunk 1]
```diff
@@ -170,6 +170,96 @@ func (q Querier) PoolParams(ctx context.Context, req *types.QueryPoolParamsReque
...
 
+// PoolsWithFilter query allows to query pools with specific parameters
+func (q Querier) PoolsWithFilter(ctx context.Context, req *types.QueryPoolsWithFilterRequest) (*types.QueryPoolsWithFilterResponse, error) {
+	res, err := q.Pools(ctx, &types.QueryPoolsRequest{
+		Pagination: &query.PageRequest{},
+	})
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	if err != nil {
+		panic(err)
```
Comment (mattverse) [inline x/gamm/keeper/grpc_query.go:181]: this should return error, not panic!

File: x/gamm/keeper/grpc_query_test.go [hunk 2]
```diff
@@ -87,6 +89,143 @@ func (suite *KeeperTestSuite) TestQueryPools() {
...
+		}
+	)
+
+	testCases := []struct {
+		name               string
+		num_pools          int
+		expected_num_pools int
+		min_liquidity      sdk.Coins
+		pool_type          string
+		include_denom      string
```
Comment (mattverse) [inline x/gamm/keeper/grpc_query_test.go:112]:
```suggestion
```

unused field

File: x/gamm/keeper/grpc_query_test.go [hunk 3]
```diff
@@ -87,6 +89,143 @@ func (suite *KeeperTestSuite) TestQueryPools() {
...
+		poolAssets         []balancertypes.PoolAsset
+		expectedErr        error
+	}{
+		{
+			name:               "valid tc",
```
Comment (mattverse) [inline x/gamm/keeper/grpc_query_test.go:117]:
```suggestion
			name:               "valid tc with both filters for min liquidity and pool type",
```

File: x/gamm/keeper/grpc_query_test.go [hunk 4]
```diff
@@ -87,6 +89,143 @@ func (suite *KeeperTestSuite) TestQueryPools() {
...
+			num_pools:          2,
```
Comment (mattverse) [inline x/gamm/keeper/grpc_query_test.go:118]: can we add two test cases for:
- only querying using filter for min liquidity
- only querying using fitler for pool type

**Review by mattverse — COMMENTED** [review]
Left some initial comments for this PR! Looking Good! Please take a look!

File: x/gamm/keeper/grpc_query.go [hunk 5]
```diff
@@ -170,6 +170,96 @@ func (q Querier) PoolParams(ctx context.Context, req *types.QueryPoolParamsReque
...
+	}
+	pools := res.Pools
+
+	var response = []*codectypes.Any{}
+
+	// set filters
+	min_liquidity := req.MinLiquidity
+	pool_type := req.PoolType
+	checks_needed := 0
```
[the remark this answers is not in the document]
Comment (RusAkh) [reply x/gamm/keeper/grpc_query.go:190]: I do not think we can do that:
```
if min_liquidity_flag || pool_type_flag {
			response = append(response, p)
}
```
I just tried to do something like that and as you can see this logic will not work, we still need to somehow know how many checks we need.

**Comment by RusAkh** [comment]
hello @mattverse done

**Review by mattverse — APPROVED** [review]
LGTM, @RusAkh  can you -help with resolving the final merge conflict and then we can merge

**Comment by RusAkh** [comment]
hello @mattverse I think it is done

**Comment by ValarDragon** [comment]
I haven't reviewed this, but I'm happy to merge as it only adds a new queries

**Comment by jonator** [comment]
Can we eventually create a query that will calculate spot prices and compare them amongst pools? That would make things much simpler on frontend. Then, we'd be able to just supply one amount, say OSMO or USDC, and the query would calculate and only return pools with a minimum equivalent amount of the given asset. Then, we could just supply 100,000 USDC and the query would return pools with a TVL of at least $100,000

**Comment by RusAkh** [comment]
@jonator If I understand you correctly, initially I had this idea of `minimum liquidity` filter, however it was rejected. Is it what you mean, right?

If so, I would like to work on it and bring back the filter if @mattverse approves

**Comment by RusAkh** [comment]
Messages from this PR:
```
currently min_liquidity is used in a format of sdk coins such that an amount of 
every token in pool should be greater or equal than each sdk coin in min_liquidity request.
I thought about using an actual price for liquidity in this query (for example 1 mil $) 
but do not really know how to implement this. 
What do you think should be done here?
```
```
So I don't think the sdk coin amt represents liquidity. I propose we drop the min_liquidity feature here
```

I think the actual liquidity feature could be implemented on an application level using oracles and in this query we could leave `min_liquidity` as `sdk.Coins`. Even so, I guess an actual price$ won't be always necessary for a user. I think, it is also good to be able to specify the number of tokens the pool in response would need to contain in spite of their actual price

**Pull request merged**